src/embeddedTraceViewer.ts

```diff
@@ -0,0 +1,260 @@
+  }
+
+  private async _computeServerUrlPrefix() {
+    if (!this._testServer)
+      return;
+    const wsEndpoint = await this._testServer.wsEndpoint();
+    if (!wsEndpoint)
+      return;
+    const serverUrl = new URL(wsEndpoint);
+    serverUrl.protocol = 'http';
+    if (this.vscode.env.remoteName && ['[::1]', '0.0.0.0'].includes(serverUrl.hostname))
+      serverUrl.hostname = 'localhost';
+    const serverUri = await this.vscode.env.asExternalUri(this.vscode.Uri.parse(serverUrl.origin));
+    return serverUri.toString().replace(/\/$/, '');
+  }
+
+  private async _startIfNeeded() {
+    if (this._traceViewerPanel)
+      return;
+    const serverUrlPrefix = await this._computeServerUrlPrefix();
+    if (!serverUrlPrefix)
+      return;
+    this._traceViewerPanel = new EmbeddedTraceViewerPanel(this, serverUrlPrefix);
+  }
+
+  checkSupport(userGesture?: boolean) {
+    if (!this._settingsModel.showTrace.get() || !this._settingsModel.embeddedTraceViewer.get())
+      return false;
+
+    const version = 1.46;
+    if (this._config.version < version) {
+      if (userGesture) {
+        const featureName = this.vscode.l10n.t('embedded trace viewer');
+        this.vscode.window.showWarningMessage(
+            this.vscode.l10n.t('Playwright v{0}+ is required for {1} to work, v{2} found', version, featureName, this._config.version)
+        );
+      }
+      return false;
+    }
+    return true;
+  }
+
+  infoForTest() {
+    if (!this._traceViewerPanel)
+      return;
+    return {
+      type: 'embedded',
+      serverUrlPrefix: this._traceViewerPanel?.serverUrlPrefix,
+      testConfigFile: this._config.configFile,
+      traceFile: this.currentFile(),
+    };
+  }
+}
+
+class EmbeddedTraceViewerPanel extends DisposableBase {
+
+  public static readonly viewType = 'playwright.traceviewer.view';
+
+  private _vscode: vscodeTypes.VSCode;
+  private _extensionUri: vscodeTypes.Uri;
+  private _webviewPanel: vscodeTypes.WebviewPanel;
+  readonly serverUrlPrefix: string;
+
+  constructor(
+    embeddedTestViewer: EmbeddedTraceViewer,
+    serverUrlPrefix: string
+  ) {
+    super();
+    this._vscode = embeddedTestViewer.vscode;
+    this._extensionUri = embeddedTestViewer.extensionUri;
+    this.serverUrlPrefix = serverUrlPrefix;
+    this._webviewPanel = this._vscode.window.createWebviewPanel(EmbeddedTraceViewerPanel.viewType, 'Trace Viewer', {
+      viewColumn: this._vscode.ViewColumn.Active,
+      preserveFocus: true,
+    }, {
+      enableScripts: true,
+      enableForms: true,
+    });
+    this._webviewPanel.iconPath = this._vscode.Uri.joinPath(this._extensionUri, 'images', 'playwright-logo.svg');
+    this._webviewPanel.webview.html = this._getHtml();
+    this._disposables = [
+      this._webviewPanel,
+      this._webviewPanel.onDidDispose(() => {
+        embeddedTestViewer.close();
+      }),
+      this._webviewPanel.webview.onDidReceiveMessage(({ command, params }) => {
+        if (typeof command === 'string')
```

Comment:
What else are we expecting? Is it safe to destructure it one line above?

Reply:
Messages that get here are from `playwright/packages/trace-viewer` module.

I was reviewing the changes I did there and unfortunatelly I used different message keys, both `command` and `message`:

- `command: 'openExternal'`: 

https://github.com/microsoft/playwright/blob/7ec3a93db3607e39b001e700020635bb017b4508/packages/trace-viewer/src/ui/embeddedWorkbenchLoader.tsx#L26

- `method: 'showErrorMessage'`:

https://github.com/microsoft/playwright/blob/7ec3a93db3607e39b001e700020635bb017b4508/packages/trace-viewer/src/ui/embeddedWorkbenchLoader.tsx#L82

So I need to change this, and I'll also create a new PR in playwright to make them consistent (with `method`, which is already very common in playwright code for these kind of messages).